database/database.go

```diff
@@ -383,6 +379,14 @@ func (db *Impl) SaveCommitSignatures(ctx context.Context, signatures []*types.Co
 }
 
 func (db *Impl) SaveBucket(ctx context.Context, bucket *models.Bucket) error {
+	txAny := ctx.Value(common.UseTx(UseTransaction))
+	if tx, ok := txAny.(*Impl); ok && tx != nil {
```

Comment:
a little weird, why not just set a bool value in context

Reply:
If it is bool, how do I get the db instance that started the transaction

Reply:
the func receiver?

Reply:
If func receives, transactions are not required in all scenarios, so consider using the context tape

Reply:
func (db *Impl) SaveBucket(ctx context.Context, bucket *models.Bucket) error {
	useTx := ctx.Value("use_transaction")
	tx := db
	if useTransaction, ok := useTx.(bool); ok && useTransaction {
		tx = tx.Begin(ctx)
	}

	err := tx.Db.Table((&models.Bucket{}).TableName()).Clauses(clause.OnConflict{
		Columns:   []clause.Column{{Name: "bucket_id"}},
		UpdateAll: true,
	}).Create(bucket).Error
	return err
}